File: sdk/go/provider/shitty_email.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const shittyEmailBase = "https://shitty.email/api"

type shittyEmailInboxResponse struct {
	Email     string                   `json:"email"`
	Token     string                   `json:"token"`
	ExpiresAt string                   `json:"expiresAt"`
	Emails    []map[string]interface{} `json:"emails"`
}

func shittyEmailJSON(method, u, token string, out any) error {
	req, err := http.NewRequest(method, u, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	if token != "" {
		req.Header.Set("X-Session-Token", token)
	}
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("shitty-email: http %d", resp.StatusCode)
	}
	if out == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func shittyEmailRaw(raw map[string]interface{}, recipient string) map[string]interface{} {
	to := raw["to"]
	if to == nil || fmt.Sprint(to) == "" {
		to = recipient
	}
	return map[string]interface{}{
		"id":      raw["id"],
		"from":    raw["from"],
		"to":      to,
		"subject": raw["subject"],
		"text":    raw["text"],
		"html":    raw["html"],
		"date":    raw["date"],
	}
}

func ShittyEmailGenerate() (*CreatedMailbox, error) {
	var data shittyEmailInboxResponse
	if err := shittyEmailJSON(http.MethodPost, shittyEmailBase+"/inbox", "", &data); err != nil {
		return nil, err
	}
	email := strings.TrimSpace(data.Email)
	token := strings.TrimSpace(data.Token)
	if email == "" || !strings.Contains(email, "@") || token == "" {
		return nil, fmt.Errorf("shitty-email: invalid inbox response")
	}
	return &CreatedMailbox{
		Channel:   "shitty-email",
		Email:     email,
		Token:     token,
		ExpiresAt: data.ExpiresAt,
	}, nil
}

func shittyEmailFetchDetail(token, id string) (map[string]interface{}, error) {
	u := fmt.Sprintf("%s/email/%s", shittyEmailBase, url.PathEscape(id))
	var detail map[string]interface{}
	if err := shittyEmailJSON(http.MethodGet, u, token, &detail); err != nil {
		return nil, err
	}
	return detail, nil
}

func ShittyEmailGetEmails(token, email string) ([]NormEmail, error) {
	sessionToken := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if sessionToken == "" {
		return nil, fmt.Errorf("shitty-email: empty token")
	}
	if address == "" {
		return nil, fmt.Errorf("shitty-email: empty email")
	}

	var data shittyEmailInboxResponse
	if err := shittyEmailJSON(http.MethodGet, shittyEmailBase+"/inbox", sessionToken, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Emails))
	for _, row := range data.Emails {
		raw := row
		id := strings.TrimSpace(fmt.Sprint(row["id"]))
		if id != "" {
			if detail, err := shittyEmailFetchDetail(sessionToken, id); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(shittyEmailRaw(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/shitty-email.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'shitty-email';
const API_BASE = 'https://shitty.email/api';

interface InboxResponse {
  email?: string;
  token?: string;
  expiresAt?: string;
  emails?: Array<Record<string, unknown>>;
}

function flattenMessage(raw: Record<string, unknown>, recipient: string): Record<string, unknown> {
  return {
    id: raw.id,
    from: raw.from,
    to: raw.to || recipient,
    subject: raw.subject,
    text: raw.text,
    html: raw.html,
    date: raw.date,
  };
}

async function fetchJSON<T>(url: string, token?: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: 'application/json',
      'User-Agent': 'Mozilla/5.0',
      ...(token ? { 'X-Session-Token': token } : {}),
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`shitty-email http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await fetchJSON<InboxResponse>(`${API_BASE}/inbox`, undefined, { method: 'POST' });
  const email = String(data.email || '').trim();
  const token = String(data.token || '').trim();
  if (!email || !email.includes('@') || !token) {
    throw new Error('shitty-email: invalid inbox response');
  }
  return {
    channel: CHANNEL,
    email,
    token,
    expiresAt: data.expiresAt,
  };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const sessionToken = String(token || '').trim();
  const address = String(email || '').trim();
  if (!sessionToken) throw new Error('shitty-email: empty token');
  if (!address) throw new Error('shitty-email: empty email');

  const data = await fetchJSON<InboxResponse>(`${API_BASE}/inbox`, sessionToken);
  const rows = Array.isArray(data.emails) ? data.emails : [];
  const emails: Email[] = [];
  for (const row of rows) {
    const id = String(row.id || '').trim();
    let raw = row;
    if (id) {
      try {
        raw = await fetchJSON<Record<string, unknown>>(`${API_BASE}/email/${encodeURIComponent(id)}`, sessionToken);
      } catch {}
    }
    emails.push(normalizeEmail(flattenMessage(raw, address), address));
  }
  return emails;
}
